File: frontend/src/utils/permissions.js
// Reusable helper for reading the logged-in user's Menu Permissions
// (set in Platform Roles > Menu Permissions, resolved once at login and
// stored on the user object in localStorage as `allowed_menus`).
//
// Platform Administrator always gets full access, matching the same
// behavior already used at login and in the backend's require_permission().

const EMPTY_PERMS = {
  can_view: false,
  can_create: false,
  can_edit: false,
  can_delete: false,
  can_export: false,
  can_approve: false
};

const FULL_PERMS = {
  can_view: true,
  can_create: true,
  can_edit: true,
  can_delete: true,
  can_export: true,
  can_approve: true
};

const getStoredUser = () => {
  try {
    const saved = localStorage.getItem('ranalyzer_user');
    return saved ? JSON.parse(saved) : null;
  } catch (err) {
    console.warn('Could not read stored user for permission check:', err);
    return null;
  }
};

export const getMenuPermission = (menuName) => {
  const user = getStoredUser();
  if (!user) return EMPTY_PERMS;
  if (user.role === 'Platform Administrator') return FULL_PERMS;

  const allowedMenus = user.allowed_menus || [];
  const found = allowedMenus.find((m) => m.menu_name === menuName);
  return found || EMPTY_PERMS;
};

export const canView = (menuName) => getMenuPermission(menuName).can_view;
export const canCreate = (menuName) => getMenuPermission(menuName).can_create;
export const canEdit = (menuName) => getMenuPermission(menuName).can_edit;
export const canDelete = (menuName) => getMenuPermission(menuName).can_delete;
export const canExport = (menuName) => getMenuPermission(menuName).can_export;
export const canApprove = (menuName) => getMenuPermission(menuName).can_approve;

File: backend/app/utils/permissions.py
from fastapi import Header, HTTPException, Depends, status
from sqlalchemy.orm import Session
from sqlalchemy import or_

from app.database import get_db
from app.models.platform_role import PlatformRole
from app.models.menu_permission import MenuPermission

ACTION_FIELDS = {
    "view": "can_view",
    "create": "can_create",
    "edit": "can_edit",
    "delete": "can_delete",
    "export": "can_export",
    "approve": "can_approve",
}


def require_permission(menu_name: str, action: str):
    """
    Returns a FastAPI dependency that blocks the request with 403 unless the
    calling user's role has the given action enabled for the given menu in
    the Menu Permissions table (Platform Roles > Menu Permissions).

    Platform Administrator always passes, matching the same full-access
    behavior already used at login and throughout menu permission seeding.

    Relies on the X-User-Role header, which the frontend already attaches
    to every request (see dashboardService.js apiClient interceptor).
    """
    field = ACTION_FIELDS.get(action, "can_view")

    def dependency(
        x_user_role: str = Header(default=None),
        db: Session = Depends(get_db)
    ):
        if x_user_role == "Platform Administrator":
            return True

        if not x_user_role:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=f"Access denied: no role context provided for '{menu_name}'."
            )

        role = db.query(PlatformRole).filter(
            or_(
                PlatformRole.role_name == x_user_role,
                PlatformRole.role_code == x_user_role
            )
        ).first()

        if not role:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=f"Access denied: role '{x_user_role}' is not recognized."
            )

        perm = db.query(MenuPermission).filter(
            MenuPermission.role_id == role.id,
            MenuPermission.menu_name == menu_name
        ).first()

        allowed = getattr(perm, field, False) if perm else False
        if not allowed:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=f"Your role ('{x_user_role}') does not have {action} permission for {menu_name}."
            )

        return True

    return dependency
